### PR TITLE
Oliver

### PR DESCRIPTION
**Pull request by AdamTegelberg**

### CODE REVIEW

File: lab4.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import java.util.Scanner;  
+
+
+public class lab4 {
+
+
+    public static void main(String[] args) {
+
+        Scanner scanner = new Scanner(System.in);
```
Comment (emmahaggren) [inline lab4.java]: Unneccessary line break.
```suggestion
        Scanner scanner = new Scanner(System.in);
```

File: lab4.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        System.out.println("search <pattern> <file>");
+
+        String userInput = scanner.nextLine();
+        String pattern = userInput.split("\\s+")[1];
+        String fileName = userInput.split("\\s+")[2]; 
+
+        
+
```
Comment (emmahaggren) [inline lab4.java]: Too many line breaks
```suggestion

```

File: lab4.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        TextReader reader = new TextReader();
+        reader.read(fileName, pattern);
+
+        scanner.close();
+        
+
+
```
Comment (emmahaggren) [inline lab4.java]: Unnecessary line breaks 
```suggestion
```

File: lab4.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        TextReader reader = new TextReader();
+        reader.read(fileName, pattern);
+
+        scanner.close();
+        
+
+        
+    }
+
```
Comment (emmahaggren) [inline lab4.java]: Unnecessary line break
```suggestion
```

File: TextReader.java [hunk 5]
```diff
@@ -1 +1,79 @@
...
+import java.util.Scanner;
+import java.util.Set;
 
+public class TextReader {
+
+	private ArrayList<String> lines = new ArrayList<String>();
+	Map<String, ArrayList<Integer>> wordMap = new HashMap<>();
+
+	public void read(String fileName, String pattern) {
+
```
Comment (emmahaggren) [inline TextReader.java]: Unnecessary line break
```suggestion
```

File: TextReader.java [hunk 6]
```diff
@@ -1 +1,79 @@
...
+		setUpWordMap(fileName);
+
+		Set<Integer> matchedLines = matchWords(wordMap, pattern);
+
+		printResults(matchedLines,lines);
```
Comment (emmahaggren) [inline TextReader.java]: Added space between "," and "lines"
```suggestion
		printResults(matchedLines, lines);
```

File: TextReader.java [hunk 7]
```diff
@@ -1 +1,79 @@
...
+				for (String word : line.split("\\s+")) {
+					if (!wordMap.containsKey(word)) {
+						wordMap.put(word, new ArrayList<>());
+						wordMap.get(word).add(index);
+					} else {
+						wordMap.get(word).add(index);
+					}
+				}
+				index++;
+			}
```
Comment (emmahaggren) [inline TextReader.java]: Searching for "sister" gives no result, but searching for "sister." does, we suggest stripping the raw word after the split of all non-letter characters.
```suggestion
                        // adds all words in each line into wordList
                        int index = 0;
                        for (String line : lines) {
                            for (String rawWord : line.split("\\s+")) {
                                String word = rawWord.replaceAll("[^a-zA-Z]", "");
                                if (!wordMap.containsKey(word)) {
                                    wordMap.put(word.replaceAll("[^a-zA-Z]", ""), new ArrayList<>());
                                    wordMap.get(word).add(index);
                                } else {
                                    wordMap.get(word).add(index);
                                }
                            }
                            index++;
                        }
```

File: TextReader.java [hunk 8]
```diff
@@ -1 +1,79 @@
...
+		}
+		
+	}
+
+	private Set<Integer> matchWords(Map<String, ArrayList<Integer>> words, String pattern) {
+    	Set<Integer> matchedLines = new HashSet<Integer>();
+		for(Map.Entry<String, ArrayList<Integer>> entry : words.entrySet() ) {
+			if (entry.getKey().trim().equals(pattern.trim())) {
+				for (int index : entry.getValue()) {
+					matchedLines.add(index);
```
Comment (emmahaggren) [inline TextReader.java]: We noticed that we always got the line after the line with our word, so we suggest changing the indexing to fix this bug
```suggestion
					matchedLines.add(index-1);
```

**Review by emmahaggren — CHANGES_REQUESTED** [review]

**Review by oliverlevay — APPROVED** [review]

**Review by emmahaggren — APPROVED** [review]
LGTMAO!

**Review by krisedtemakatt — COMMENTED** [review]
nice work

**Pull request merged**